workers/unity/Packages/io.improbable.gdk.core/.codegen/Source/Generators/Core/UnityComponentReplicationSystemGenerator.cs

```diff
@@ -0,0 +1,98 @@
+
+namespace Improbable.Gdk.CodeGenerator
+{
+    public static class UnityComponentReplicationSystemGenerator
+    {
+        private static readonly Logger Logger = LogManager.GetCurrentClassLogger();
+
+        public static CodeWriter Generate(UnityComponentDetails componentDetails)
+        {
+            var componentNamespace = $"global::{componentDetails.Namespace}.{componentDetails.Name}";
+
+            Logger.Trace($"Generating {componentDetails.Namespace}.{componentDetails.Name}.ReplicationSystem internal class.");
+
+            return CodeWriter.Populate(cgw =>
+            {
+                cgw.UsingDirectives(
+                    "Improbable.Gdk.Core",
+                    "Improbable.Worker.CInterop",
+                    "Unity.Collections",
+                    "Unity.Entities",
+                    "Unity.Profiling"
+                );
+
+                cgw.Namespace(componentDetails.Namespace, ns =>
+                {
+                    ns.Type($"public partial class {componentDetails.Name}", partial =>
+                    {
+                        partial.Annotate("DisableAutoCreation, UpdateInGroup(typeof(SpatialOSSendGroup)), UpdateBefore(typeof(SpatialOSSendGroup.InternalSpatialOSSendGroup))")
+                            .Type("internal class ReplicationSystem : SystemBase", system =>
+                            {
+                                system.Line($@"
+private NativeQueue<SerializedMessagesToSend.UpdateToSend> dirtyComponents;
+private SpatialOSSendSystem spatialOsSendSystem;
+
+private ProfilerMarker foreachMarker = new ProfilerMarker(""{componentDetails.Name}SerializationJob"");
+
+protected override void OnCreate()
+{{
+    spatialOsSendSystem = World.GetExistingSystem<SpatialOSSendSystem>();
+}}
+");
+                                system.Method("protected override void OnUpdate()", m =>
+                                {
+                                    m.Line(new[]
+                                    {
+                                        "dirtyComponents = new NativeQueue<SerializedMessagesToSend.UpdateToSend>(Allocator.TempJob);",
+                                        "var dirtyComponentsWriter = dirtyComponents.AsParallelWriter();",
+                                        "var marker = foreachMarker;",
+                                    });
+
+                                    m.Line($@"
+Dependency = Entities.WithName(""{componentDetails.Name}Replication"")");
+                                    if (!componentDetails.IsBlittable)
+                                    {
+                                        m.Line(@"
+    .WithoutBurst()");
+                                    }
+
+                                    m.Line(@"
+    .WithAll<HasAuthority>()
+    .WithChangeFilter<Component>()
+    .ForEach((ref Component component, in SpatialEntityId entity) =>
+    {
+        marker.Begin();
+        if (!component.IsDataDirty())
+        {
+            marker.End();
```

Comment:
Is there a scoped version of the marker rather than handling the various branches ourselves?

Reply:
Yes, but this can be bursted, and our version of burst does not yet support `using`

Reply:
Ah right, soon though 😉